Replace `inctimer` package with `time.After`

tklauser opened the pull request:
Go version ≥ 1.23 no longer has the issue of not garbage-collecting unstopped
`time.Timer`s and thus `time.After` can again safely be used in loops. Also see
https://go.dev/doc/go1.23#timer-changes and
https://cs.opensource.google/go/go/+/refs/tags/go1.23.2:src/time/sleep.go;l=196-201 for details.

This PR pulls in the linter change which skips the corresponding `timeafter` check on Go version ≥ 1.23[^1], converts all uses of the `inctimer` package back to `time.After` and removes the `inctimer` package.

See individual commit messages for details.

[^1]: https://github.com/cilium/linters/pull/48

tklauser committed:
go.mod: bump github.com/cilium/linters to v0.1.0

This is the first tagged version and includes support for skipping the
timeafter check on Go version ≥ 1.23 [1] which is a prerequisite to
dropping the inctimer package.

[1] https://github.com/cilium/linters/pull/48

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
cilium-cli: drop use of pkg/inctimer

Go version ≥ 1.23 no longer has the issue of not collecting unstopped
Timers and time.After can safely be used in loops. Also see
https://go.dev/doc/go1.23#timer-changes and
https://cs.opensource.google/go/go/+/refs/tags/go1.23.2:src/time/sleep.go;l=196-201

Replace any (*incTimer.After) and inctimer.After calls with time.After
accordingly.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
all: drop use of pkg/inctimer

Go version ≥ 1.23 no longer has the issue of not collecting unstopped
Timers and time.After can safely be used in loops. Also see
https://go.dev/doc/go1.23#timer-changes and
https://cs.opensource.google/go/go/+/refs/tags/go1.23.2:src/time/sleep.go;l=196-201

Replace any (*incTimer.After) and inctimer.After calls with time.After
accordingly.

Also note that this change also makes sure that maximum sleepiness is
enforced consistently via option.MaxInternalTimerDelay in pkg/time if
needed during testing. This wasn't the case before when using
inctimer.After which directly used stdlib time.After.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

giorio94 reviewed (approved):
Thanks! Looks good to me for my codeowners.

tklauser committed:
inctimer: remove package

Now that all users of the package are migrated to use time.After again
(see previous commits), the package is no longer used and can be
removed.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser commented on pkg/controller/controller.go:
```diff
@@ -325,7 +322,7 @@ func (c *controller) runController(params ControllerParams) {
 		case params = <-c.update:
 			// update channel is never closed
-		case <-runTimer.After(interval):
+		case <-stdtime.After(interval):
```
Note that we have to use stdlib `time` here because in the successful case `interval` is set to `math.MaxInt64` to wait for the next controller update or stop but that would be overridden by `pkg/time.After` which would break controller semantics.

joestringer replied:
Will this render the change in https://github.com/cilium/cilium/pull/27253 ineffective?

joestringer replied:
Ah I guess that `inctimer` wasn't using `pkg/time` anyway so this doesn't change anything in that regard :+1: . I didn't fully understand the implication here yet.. worth thinking about but this doesn't seem to break anything.

tklauser replied:
> Will this render the change in https://github.com/cilium/cilium/pull/27253 ineffective?

Yes, this will render that change ineffective as was previously the case due to using `inctimer.After` which didn't use `pkg/time`. I don't fully understand the implications too, but I saw Ginkgo tests failing reliably in CI when using `pkg/time.After` here. It seems Ginkgo tests set `--max-internal-timer-delay=5s` which will cap the interval to 5s instead of `mat.MaxInt64` which the `controller` logic seems to require for its semantics to work. I'd assume we'd need to rework how `controller` works in case of a successful `DoFunc` if we want to start using `pkg/time` here too but I'd say that's out of scope for this PR.

joestringer replied:
I'm thinking now that this is just one additional case that #27253 did not fully consider. For controllers that only run once, it doesn't make sense to reset the interval to `pkg/time:MaxInternalTimerDelay` as this would break controller semantics, I agree. On the other hand, for controllers that regularly run it could be useful to ensure controllers run more frequently in CI in order to catch potential "eventual consistency issues" where subsequent runs of a controller might break something.

Anyways I think we can iterate on that further outside of this PR, it's not directly relevant to what you're trying to achieve.

tklauser replied:
FWIW, there was a similar case in `pkg/maps/ctmap/gc/gc.go` (i.e. previous use of `inctimer.After` replaced by `stdtime.After`) which also lead to reliable test failures when switching to `pkg/time`.

joestringer replied:
Yeah I noticed that one too. I'll add some tasks to https://github.com/cilium/cilium/issues/28844 to investigate further.

rolinh reviewed (approved):
Thanks! Hubble changes lgtm.
